skyportal/models/phot_stat.py

```diff
@@ -0,0 +1,668 @@
+                filt_mags = good_mags[good_filters == filt]
+
+                if len(filt_mjds):
+                    self.num_det_per_filter[filt] = len(filt_mjds)
+                    self.mean_mag_per_filter[filt] = np.nanmean(filt_mags)
+                    self.mag_rms_per_filter[filt] = np.nanstd(filt_mags)
+                    self.peak_mag_per_filter[filt] = min(filt_mags)
+                    self.faintest_mag_per_filter[filt] = max(filt_mags)
+
+            # find all the color terms
+            mean_mags = self.mean_mag_per_filter
+            for f1 in set(good_filters):
+                for f2 in set(good_filters):
+                    if f1 != f2:
+                        self.mean_color[f'{f1}-{f2}'] = mean_mags[f1] - mean_mags[f2]
+
+        # if any are non-detections
+        if np.any(dets == 0):
+            lim_mags = lims[dets == 0]
+            lim_mjds = mjds[dets == 0]
+            lim_filters = filters[dets == 0]
+            # find the deepest limit
+            self.deepest_limit_global = max(lim_mags)
+            if self.first_detected_mjd:
+                mjds_before = lim_mjds[lim_mjds < self.first_detected_mjd]
+            else:
+                mjds_before = lim_mjds
+
+            if len(mjds_before):
+                self.last_non_detection_mjd = max(mjds_before)
+                if self.first_detected_mjd and self.last_non_detection_mjd:
+                    self.time_to_non_detection = (
+                        self.first_detected_mjd - self.last_non_detection_mjd
+                    )
+
+            # stats for non-detections for each filter
+            for filt in set(lim_filters):
+                filt_mags = lim_mags[lim_filters == filt]
+                if len(filt_mags):
+                    self.deepest_limit_per_filter[filt] = max(filt_mags)
+
+        self.last_update = datetime.utcnow()
+        self.last_full_update = datetime.utcnow()
+
+    @staticmethod
+    def update_average(current, number, new):
+        """
+        Calculate the new average given the current average,
+        the number of points used to calculate that average,
+        and the new value to be added.
+        """
+        return (number * current + new) / (number + 1)
+
+    @staticmethod
+    def update_scatter(current_scatter, current_mean, number, new):
+        """
+        Calculate the new scatter (RMS) given the current scatter,
+        the current average,
+        the number of points used to calculate the average/scatter,
+        and the new value to be added.
+        """
+        var = current_scatter**2
+
+        new_var = (
+            number / (number + 1) * (var + ((current_mean - new) ** 2) / (number + 1))
+        )
+
+        return np.sqrt(new_var)
+
+
+@event.listens_for(Photometry, 'after_insert')
+def insert_into_phot_stat(mapper, connection, target):
+
+    # Create or update PhotStat object
+    @event.listens_for(DBSession(), "before_flush", once=True)
```

Comment:
why is this before_flush?

Reply:
Before flush has less options for what you can change in the session state, and I could not get it to work at all.